janus_server/src/datastore.rs

```diff
@@ -0,0 +1,318 @@
+            VALUES ($1, $2, $3, $4, $5) RETURNING (id)",
+            &[
+                /* task_id */      &&report.task_id.0[..],
+                /* nonce_time */   &nonce_time,
+                /* nonce_rand */   &nonce_rand,
+                /* extensions */   &encoded_extensions,
+                /* input_shares */ &encoded_input_shares,
+            ]
+        ).await?;
+        Ok(row.get("id"))
+    }
+}
+
+fn single_row(rows: Vec<Row>) -> Result<Row, Error> {
+    match rows.len() {
+        0 => Err(Error::NotFound),
+        1 => Ok(rows.into_iter().next().unwrap()),
+        _ => Err(Error::TooManyRows),
+    }
+}
+
+/// Error represents a datastore-level error.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// An error from the underlying database library.
+    #[error("DB error: {0}")]
+    Db(#[from] tokio_postgres::Error),
+    /// An entity requested from the datastore was not found.
+    #[error("not found in database")]
+    NotFound,
+    /// A query that was expected to return at most one row unexpectedly returned more than one row.
+    #[error("multiple rows returned where only one row expected")]
+    TooManyRows,
+    /// An error from decoding a value stored encoded in the underlying database.
+    #[error("decoding error: {0}")]
+    DecodeError(#[from] CodecError),
+    /// An arbitrary error returned from the user callback; unrelated to DB internals. This error
+    /// will never be generated by the datastore library itself.
+    #[error(transparent)]
+    User(#[from] Box<dyn std::error::Error + Send + Sync>),
+}
+
+impl Error {
+    // is_serialization_failure determines if a given error corresponds to a Postgres
+    // "serialization" failure, which requires the entire transaction to be aborted & retried from
+    // the beginning per https://www.postgresql.org/docs/9.5/transaction-iso.html.
+    fn is_serialization_failure(&self) -> bool {
+        match self {
+            // T_R_SERIALIZATION_FAILURE (40001) is documented as the error code which is always used
+            // for serialization failures which require rollback-and-retry.
+            Error::Db(err) => err
+                .code()
+                .map_or(false, |c| c == &SqlState::T_R_SERIALIZATION_FAILURE),
+            _ => false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    // TODO(brandon): use podman instead of docker for container management once testcontainers supports this
+
+    use crate::message::{ExtensionType, HpkeConfigId};
+
+    use super::*;
+    use lazy_static::lazy_static;
+    use std::str;
+    use testcontainers::{clients, images::postgres::Postgres, Container, Docker};
+    use tokio_postgres::NoTls;
+
+    lazy_static! {
+        static ref SCHEMA: &'static str = {
+            let schema_bytes = include_bytes!("../../db/schema.sql");
+            str::from_utf8(schema_bytes).unwrap()
+        };
+    }
```

Comment:
(will also require dropping a * below)
```suggestion
    static SCHEMA: &'static str = include_str!("../../db/schema.sql");
```